### PR TITLE
Add dev docs for examples verify.sh

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: Add dev docs for examples verify.sh
Additional Description:
Adds docs for adding example sandboxes

Documentation for #12491  - separated here so as not to complexify that pr any further.

Risk Level: low
Testing:
Docs Changes: yep
Release Notes:
[Optional Runtime guard:]
[Optional Fixes #Issue] related to #12491 
[Optional Deprecated:]

### CODE REVIEW

File: examples/DEVELOPER.md [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+
+Sandboxes are tested as part of the continous integration process, which expects
+each sandbox to have a `verify.sh` script containing tests for the example.
+
+### Basic layout of the `verify.sh` script
+
+At a minimum the verify.sh script should include the necessary parts to start
+and stop your sandbox.
+
+Given a sandbox with single a single `docker` composition, adding the following
```
Comment (phlax) [inline examples/DEVELOPER.md:13]: repeat

File: examples/DEVELOPER.md [hunk 2]
```diff
@@ -0,0 +1,179 @@
+# Adding a sandbox example
+
+## Add a `verify.sh` test to your sandbox
+
+Sandboxes are tested as part of the continous integration process, which expects
+each sandbox to have a `verify.sh` script containing tests for the example.
+
+### Basic layout of the `verify.sh` script
+
+At a minimum the verify.sh script should include the necessary parts to start
```
Comment (phlax) [inline examples/DEVELOPER.md:10]: `verify.sh`

File: examples/DEVELOPER.md [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+The tests should follow the steps laid out in the documentation.
+
+For example, if the documentation provides a series of `bash` commands to execute, add these in order to `verify.sh`.
+
+You may wish to grep the response, or check the return code to ensure the command responds as expected.
+
+Likewise, if the documentation asks the user to browse to a page - eg http://localhost:8000 - then you should add a test
+to ensure that the given url responds as expected.
+
+If an example web page is also expected to make further javascript http requests to function, then add tests for requests
```
Comment (phlax) [inline examples/DEVELOPER.md:55]: ...in order to function, ...

File: examples/DEVELOPER.md [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+A number of utility functions have been added to simplify browser testing.
+
+#### Utility functions: `responds_with`
+
+This can be called to ensure that the given url returns with expected http content:
+
+It follows the form `responds_with <expected_content> <url> <curl args>`
+
+An simple example `GET` request:
```
Comment (phlax) [inline examples/DEVELOPER.md:66]: For example, a ~An~

File: examples/DEVELOPER.md [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+
+```
+responds_with \
+    "Hello, world" \
+	http://localhost:8000
```
Comment (phlax) [inline examples/DEVELOPER.md:71]: untabify

File: examples/DEVELOPER.md [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+```
+
+An example of a `POST` request that expects "Hello, postie" in the returned html,
+and also sets the headers and uses `https`:
+
```
Comment (phlax) [inline examples/DEVELOPER.md:76]: This is a more complicated example that uses an `https` `POST` request and sends some additional headers:

File: examples/DEVELOPER.md [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+responds_with_header \
+    "HTTP/1.1 403 Forbidden" \
+	"http://localhost:8000/?name=notdown"
+```
+
+`responds_with_header` can accept additional curl arguments like  `responds_with`
+
+#### Utility functions: `responds_without_header`
+
+You can aslo check that a request *does not* respond with a given header:
```
Comment (phlax) [inline examples/DEVELOPER.md:101]: aslo

File: examples/DEVELOPER.md [hunk 8]
```diff
@@ -0,0 +1,179 @@
...
+    "X-Secret: treasure" \
+	"http://localhost:8000"
+```
+
+`responds_without_header` can accept additional curl arguments like  `responds_with`
+
+### Slow starting `docker` compositions
+
+Unless your example provides a way for ensuring that all containers are healthy by
+the time `docker-compose up -d` returns you may need to add a `DELAY` before running
```
Comment (phlax) [inline examples/DEVELOPER.md:114]: ...returns, you...

File: examples/DEVELOPER.md [hunk 9]
```diff
@@ -0,0 +1,179 @@
...
+the steps in your `verify.sh`
+
+For example, to wait 10 seconds after `docker-compose` has been called set the following:
```
Comment (phlax) [inline examples/DEVELOPER.md:117]: `docker-compose up -d`

File: examples/DEVELOPER.md [hunk 10]
```diff
@@ -0,0 +1,179 @@
...
+. "$(dirname "${BASH_SOURCE[0]}")/../verify-common.sh"
+
+echo foo > bar.txt
+
+bring_up_example
+
+# add example tests here...
+```
+
+If your sandbox has multiple paths, `bring_up_example` will bring all up
```
Comment (phlax) [inline examples/DEVELOPER.md:179]: ...bring all of your compositions up...

**Comment by phlax** [comment]
add sections:
- additional args to docker-compose up -d
- running commands inside docker containers

File: examples/DEVELOPER.md [hunk 11]
```diff
@@ -0,0 +1,179 @@
...
+
+# add example tests here...
+```
+
+### Bringing stacks up manually
+
+You may need to bring up the stack manually, in order to run some steps beforehand.
+
+Sourcing `verify-common.sh` will always leave you in the sandbox directory, and from there
+you can use the `bring_up_stack` function.
```
Comment (phlax) [inline examples/DEVELOPER.md:159]: `bring_up_example`

File: examples/DEVELOPER.md [hunk 12]
```diff
@@ -0,0 +1,179 @@
...
+
+```
+#!/bin/bash -e
+
+export NAME=example-sandbox
+export MANUAL=true
+
+# shellcheck source=examples/verify-common.sh
+. "$(dirname "${BASH_SOURCE[0]}")/../verify-common.sh"
+
```
Comment (phlax) [inline examples/DEVELOPER.md:171]: add `run_log`

File: examples/DEVELOPER.md [hunk 13]
```diff
@@ -0,0 +1,179 @@
...
+to ensure that the given url responds as expected.
+
+If an example web page is also expected to make further javascript http requests to function, then add tests for requests
+that mimick this interaction.
+
+A number of utility functions have been added to simplify browser testing.
+
+#### Utility functions: `responds_with`
+
+This can be called to ensure that the given url returns with expected http content:
```
Comment (phlax) [inline examples/DEVELOPER.md:62]: The `responds_with` function can be used to ensure a request to a given url responds with expected http content.

File: examples/DEVELOPER.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+. "$(dirname "${BASH_SOURCE[0]}")/../verify-common.sh"
+
+# add example tests here...
+```
+
+### Running commands inside `docker` containers
+
+If your example asks the user to run commands inside containers, you can
+mimick this using `docker-compose exec -T`. The `-T` flag is necessary as the
+tests do not have access to a `tty` in the ci pipeline.
```
Comment (dio) [inline examples/DEVELOPER.md:210]: s/ci/CI

File: examples/DEVELOPER.md [hunk 15]
```diff
@@ -0,0 +1,210 @@
...
+
+run_log "Creating bar.txt before starting containers"
+echo foo > bar.txt
+
+bring_up_example
+
+# add example tests here...
+```
+
+If your sandbox has multiple `$PATHS`, `bring_up_example` will bring all
```
Comment (dio) [inline examples/DEVELOPER.md:182]: Could you clarify what multiple `$PATH`s is?

File: examples/DEVELOPER.md [hunk 16]
```diff
@@ -0,0 +1,210 @@
...
+
+### Basic layout of the `verify.sh` script
+
+At a minimum the `verify.sh` script should include the necessary parts to start
+and stop your sandbox.
+
+Given a sandbox with a single `docker` composition, adding the following
+to `verify.sh` will test that the sandbox can be started and stopped.
+
+```
```
Comment (dio) [inline examples/DEVELOPER.md:16]: Do you think adding:

```
```bash 
```

could help readability?

File: examples/DEVELOPER.md [hunk 17]
```diff
@@ -0,0 +1,210 @@
...
+The tests should follow the steps laid out in the documentation.
+
+For example, if the documentation provides a series of `bash` commands to execute, add these in order to `verify.sh`.
+
+You may wish to grep the response, or check the return code to ensure the command responds as expected.
+
+Likewise, if the documentation asks the user to browse to a page - eg http://localhost:8000 - then you should add a test
+to ensure that the given url responds as expected.
+
+If an example web page is also expected to make further javascript http requests in order to function, then add
```
Comment (dio) [inline examples/DEVELOPER.md:55]: JavaScript HTTP requests?

File: examples/DEVELOPER.md [hunk 18]
```diff
@@ -0,0 +1,210 @@
...
+
+### Add tests reflecting the documented examples
+
+The tests should follow the steps laid out in the documentation.
+
+For example, if the documentation provides a series of `bash` commands to execute, add these in order to `verify.sh`.
+
+You may wish to grep the response, or check the return code to ensure the command responds as expected.
+
+Likewise, if the documentation asks the user to browse to a page - eg http://localhost:8000 - then you should add a test
```
Comment (dio) [inline examples/DEVELOPER.md:52]: e.g. or for example?

File: examples/DEVELOPER.md [hunk 19]
```diff
@@ -0,0 +1,210 @@
...
+to ensure that the given url responds as expected.
+
+If an example web page is also expected to make further javascript http requests in order to function, then add
+tests for requests that mimick this interaction.
+
+A number of utility functions have been added to simplify browser testing.
+
+#### Utility functions: `responds_with`
+
+The `responds_with` function can be used to ensure a request to a given url responds with
```
Comment (dio) [inline examples/DEVELOPER.md:62]: s/url/URL?

**Review by dio — COMMENTED** [review]
Looks good. Some questions:

File: examples/DEVELOPER.md (hunk 18 above)
Comment (phlax) [reply examples/DEVELOPER.md:52]: not sure that e.g. is really more correct any more - there also seems to be a localisation angle on how its used. Personally i dont bother with dots for common abbreviations. Ill switch to "for example" to lessen cultural bias

File: examples/DEVELOPER.md [hunk 20]
```diff
@@ -0,0 +1,210 @@
...
+The tests should follow the steps laid out in the documentation.
+
+For example, if the documentation provides a series of `bash` commands to execute, add these in order to `verify.sh`.
+
+You may wish to grep the responses, or check return codes to ensure the commands respond as expected.
+
+Likewise, if the documentation asks the user to browse to a page - for example http://localhost:8000 -
+then you should add a test to ensure that the given URL responds as expected.
+
+If an example web page is also expected to make further Javascript `HTTP` requests in order to function, then add
```
Comment (dio) [inline examples/DEVELOPER.md:55]: Not sure if you want to follow up fixing this: s/Javascript/JavaScript https://developer.mozilla.org/en-US/docs/Web/JavaScript 😅

**Review by dio — APPROVED** [review]
Thank you!

**Review by dio — APPROVED** [review]

**Comment by phlax** [comment]
@lizan not sure if you want to check over this before its merged

File: examples/DEVELOPER.md [hunk 21]
```diff
@@ -0,0 +1,210 @@
...
+
+If an example web page is also expected to make further JavaScript `HTTP` requests in order to function, then add
+tests for requests that mimick this interaction.
+
+A number of utility functions have been added to simplify browser testing.
+
+#### Utility functions: `responds_with`
+
+The `responds_with` function can be used to ensure a request to a given URL responds with
+expected http content.
```
Comment (phlax) [inline examples/DEVELOPER.md:63]: `HTTP`

File: examples/DEVELOPER.md [hunk 22]
```diff
@@ -0,0 +1,210 @@
...
+
+### Examples with multiple `docker` compositions
+
+For your example to work it may need more than one `docker` composition to be run.
+
+You can set where to find the `docker-compose.yml` files with the `PATHS` argument.
+
+By default `PATHS=.`, but you can change this to a comma-separated list of paths.
+
+For example a sandbox containing `frontend/docker-compose.yml` and `backend/docker-compose.yml`,
```
Comment (phlax) [inline examples/DEVELOPER.md:141]: use `docker-compose.yaml` (ie ~.yml~) throughout

**Review by dio — APPROVED** [review]
Thanks!

**Review by mattklein123 — APPROVED** [review]
Thank you so much!

**Pull request merged**